visualization/src/point_picking_event.cpp

```diff
@@ -167,49 +189,67 @@ pcl::visualization::PointPickingCallback::performSinglePick (
   vtkRenderer *ren = iren->FindPokedRenderer (iren->GetEventPosition ()[0], iren->GetEventPosition ()[1]);
   picker->AreaPick (x_, y_, iren->GetEventPosition ()[0], iren->GetEventPosition ()[1], ren);
-  if (picker->GetDataSet ())
+ if (picker->GetProp3Ds())
   {
-    vtkPolyData* points = reinterpret_cast<vtkPolyData*> (picker->GetDataSet ());
-
-    // This is a naive solution till we fugure out where to add the GlobalIds at an earlier stage
-    if (!points->GetPointData ()->GetGlobalIds () ||
-        points->GetPointData ()->GetGlobalIds ()->GetNumberOfTuples () == 0)
+    int pt_numb=0;
+    vtkProp3DCollection* props = picker->GetProp3Ds();
+    vtkCollectionSimpleIterator pit;
+    vtkProp3D* prop;
+    for(props->InitTraversal(pit);(prop = props->GetNextProp3D(pit));)
     {
-      vtkSmartPointer<vtkIdTypeArray> global_ids = vtkIdTypeArray::New ();
-      global_ids->SetNumberOfValues (picker->GetDataSet ()->GetNumberOfPoints ());
-      for (vtkIdType i = 0; i < picker->GetDataSet ()->GetNumberOfPoints (); ++i)
-        global_ids->SetValue (i,i);
+      vtkActor* actor = vtkActor::SafeDownCast(prop);
+      if(actor)
+      {
+        actors_->AddItem(actor);
+
+        std::vector<int> actorIndices;
+        vtkPolyData* pd = vtkPolyData::SafeDownCast(actor->GetMapper()->GetInput());
+
+        if(pd->GetPointData()->HasArray("Indices"))
+            pd->GetPointData()->RemoveArray("Indices");
 
-      points->GetPointData ()->SetGlobalIds (global_ids);
-    }
 
-    vtkPlanes* frustum = picker->GetFrustum ();
+        vtkSmartPointer<vtkIdTypeArray> IDs = vtkSmartPointer<vtkIdTypeArray>::New();
+        IDs->SetNumberOfComponents(1);
+        IDs->SetName("Indices");
+        for(vtkIdType i = 0; i < pd->GetNumberOfPoints(); i++)
+          IDs->InsertNextValue(i);
 
-    vtkSmartPointer<vtkExtractGeometry> extract_geometry = vtkSmartPointer<vtkExtractGeometry>::New ();
-    extract_geometry->SetImplicitFunction (frustum);
+        pd->GetPointData()->AddArray(IDs);
 
-    extract_geometry->SetInputData (picker->GetDataSet ());
+        vtkPlanes* frustum = picker->GetFrustum();
 
-    extract_geometry->Update ();
+        vtkSmartPointer<vtkExtractGeometry> extract_geometry = vtkSmartPointer<vtkExtractGeometry>::New();
+        extract_geometry->SetImplicitFunction(frustum);
+  #if VTK_MAJOR_VERSION < 6
+        extract_geometry->SetInput(pd);
+  #else
+        extract_geometry->SetInputData(pd);
+  #endif
 
-    vtkSmartPointer<vtkVertexGlyphFilter> glyph_filter = vtkSmartPointer<vtkVertexGlyphFilter>::New ();
-    glyph_filter->SetInputConnection (extract_geometry->GetOutputPort ());
-    glyph_filter->Update ();
+        extract_geometry->Update();
 
-    vtkPolyData* selected = glyph_filter->GetOutput ();
-    vtkIdTypeArray* ids = vtkIdTypeArray::SafeDownCast(selected->GetPointData ()->GetGlobalIds ());
-    assert (ids);
-    indices.reserve (ids->GetNumberOfTuples ());
```

Comment:
Any reason for removing the assert and the reserve for `actorIndices`?

Reply:
not really..

Reply:
They'd be useful, so please add them back

Reply:
resolved in new commit